src/adapter/breakpoints.ts

```diff
@@ -430,41 +477,72 @@ export class BreakpointManager {
     params.source.path = urlUtils.platformPathToPreferredCase(params.source.path);
+
+    // Wait until the breakpoint predictor finishes to be sure that we
+    // can place correctly in breakpoint.set().
     if (!this._predictorDisabledForTest && this._breakpointsPredictor) {
       const promise = this._breakpointsPredictor.predictBreakpoints(params);
       this._launchBlocker = Promise.all([this._launchBlocker, promise]);
       await promise;
     }
-    const breakpoints: Breakpoint[] = [];
-    const inBreakpoints = params.breakpoints || [];
-    for (let index = 0; index < inBreakpoints.length; index++)
-      breakpoints.push(new Breakpoint(this, ids[index], params.source, inBreakpoints[index]));
-    let previous: Breakpoint[] | undefined;
+
+    // Creates new breakpoints for the parameters, unsetting any previous
+    // breakpoints that don't still exist in the params.
+    const mergeInto = (previous: Breakpoint[]): ISetBreakpointResult => {
+      const result: ISetBreakpointResult = { unbound: previous.slice(), new: [], list: [] };
+      if (!params.breakpoints) {
+        return result;
+      }
+
+      for (let index = 0; index < params.breakpoints.length; index++) {
+        const created = new Breakpoint(this, ids[index], params.source, params.breakpoints[index]);
+        const existingIndex = result.unbound.findIndex(p => p.compare(created) === 0);
+        if (existingIndex === -1) {
+          result.new.push(created);
+          result.list.push(created);
+          continue;
+        }
+
+        const existing = result.unbound[existingIndex];
+        result.list.push(existing);
+        result.unbound.splice(existingIndex, 1);
+      }
+
+      return result;
+    };
+
+    let result: ISetBreakpointResult;
     if (params.source.path) {
-      previous = this._byPath.get(params.source.path);
-      this._byPath.set(params.source.path, breakpoints);
+      result = mergeInto(this._byPath.get(params.source.path) || []);
+      this._byPath.set(params.source.path, result.list);
     } else {
-      previous = this._byRef.get(params.source.sourceReference!);
-      this._byRef.set(params.source.sourceReference!, breakpoints);
+      result = mergeInto(this._byRef.get(params.source.sourceReference!) || []);
+      this._byRef.set(params.source.sourceReference!, result.list);
     }
+
     // Cleanup existing breakpoints before setting new ones.
-    if (previous) {
-      this._totalBreakpointsCount -= previous.length;
-      await Promise.all(previous.map(b => b.remove()));
-    }
-    this._totalBreakpointsCount += breakpoints.length;
-    if (this._thread) {
-      breakpoints.forEach(b => this._setBreakpoint(b));
+    this._totalBreakpointsCount -= result.unbound.length;
+    await Promise.all(result.unbound.map(b => b.remove()));
+
+    this._totalBreakpointsCount += result.new.length;
+
+    const thread = this._thread;
+    if (thread) {
+      await (this._launchBlocker = Promise.all([
```

Comment:
Do not await breakpoint.set() here. We report "provisionalDap" below to always notify once breakpoint is set. This way we avoid any races - always immediately report as "provisional", and then notify with real location once ready, unless the breakpoint was removed in the meantime. Also, 'notifyResolved' will be unnecessary.

Reply:
Could you describe a little more around the racing you're concerned about here? The only thing we do after this `await` is register info about the breakpoints for our own statistics, and then return the breakpoints to the UI. I think that if the UI has a race around breakpoints being set or updated before they come back from the debug adapter, then that will be a concern regardless of whether we await here or not since the call is already being made asynchronously and e.g. waits on the predictor.

Reply:
- While setting breakpoint in multiple scripts, it might already trigger in one of them. This will lead to "paused" event from adapter referencing breakpoint id which is not reported to UI just yet, because we didn't return from this method.
- Setting breakpoints might be slow, and it's nice to show "unbound" status to the user immediately until we actually verified breakpoint location.

What do you think?